File: Menu.cpp
#include "Menu.h"
#include "Utils.h"
#include <conio.h>


void Print(const std::wstring menu[], const int size)
{
    for (int i = 0; i < size; i++) {
        std::wcout << menu[i] << std::endl;
    }
    std::wcout << L"\n";
}

void BackToMenu()
{
    std::wcout << L"\n\nПожалуйста, нажмите Enter, чтобы вернуться в меню...";
    while (_getch() != 13) { continue; }
    system("cls");
}

void MainMenu(ShpaginCompany& company)
{
    static const int main_menu_size = 7;
    static const std::wstring main_menu[main_menu_size] = {
        L"1. Добавить сотрудника",
        L"2. Добавить разработчика",
        L"3. Просмотреть всех",
        L"4. Сохранить",
        L"5. Загрузить",
        L"6. Очистить",
        L"0. Выход",
    };
    Print(main_menu, main_menu_size);
    int menu = GetCorrectNumber(std::cin, 0, 6, L">> ");
    system("cls");
    switch (menu) {
    case 1:
    {
        std::wcout << L"-> Добавить сотрудника" << std::endl;
        company.add_employee();
        BackToMenu();
        break;
    }
    case 2:
    {
        std::wcout << L"-> Добавить разработчика" << std::endl;
        company.add_developer();
        BackToMenu();
        break;
    }
    case 3:
    {
        std::wcout << L"-> Просмотреть всех" << std::endl;
        std::wcout << company;
        BackToMenu();
        break;
    }
    case 4:
    {
        std::wcout << L"-> Сохранить" << std::endl;
        if (CheckBeforeSave(company)) {
            company.save_to_file();
        }
        BackToMenu();
        break;
    }
    case 5:
    {
        std::wcout << L"-> Загрузить" << std::endl;
        CheckBeforeLoad(company);
        company.load_from_file();
        BackToMenu();
        break;
    }
    case 6:
    {
        std::wcout << L"-> Очистить" << std::endl;
        company.clear_company();
        BackToMenu();
        break;
    }
    case 0:
    default:
    {
        CheckBeforeExit(company);
        exit(0);
    }
    }
}

void InputFileName(ShpaginCompany& c)
{
    c.show_saves();
    std::wcout << L"Введите имя файла: ";
    std::wstring fn;
    getline(std::wcin >> std::ws, fn);
    c.set_filename(fn);
}

bool CheckBeforeSave(ShpaginCompany& c)
{
    if (c.saved()) {
        if (c.has_saved_file()) {
            std::wcout << L"*В компании нет несохраненных изменений.\n";
            return false;
        }
        else if (c.empty()) {
            std::wcout << L"*Компания пуста\n";
            if (GetCorrectNumber(std::cin, 0, 1, L"Хотите сохранить пустой файл? (1 - да, 0 - нет): ") == 0) {
                return false;
            }
        }
    }
    else if (c.has_saved_file()) {
        int choice = GetCorrectNumber(std::cin, 0, 2, L"Хотите сохранить в том же файле? (0 - тот же, 1 - другой, 2 - не сохранять): ");
        if (choice == 2) {
            std::wcout << L"*Изменения были утеряны\n";
            return false;
        }
        if (choice == 1) {
            InputFileName(c);
        }
        return true;
    }
    else if (c.empty()) {
        std::wcout << L"*Компания пуста\n";
        if (GetCorrectNumber(std::cin, 0, 1, L"Хотите сохранить пустой файл? (1 - да, 0 - нет): ") == 0) {
            return false;
        }
    }
    InputFileName(c);
    return true;
}

void CheckBeforeExit(ShpaginCompany& c)
{
    if (!c.saved()) {
        if (c.has_saved_file()) {
            std::wcout << L"*В компании есть несохраненные изменения\n";
            int choice = GetCorrectNumber(std::cin, 0, 2, L"Хотите сохранить в том же файле? (0 - тот же, 1 - другой, 2 - не сохранять): ");
            if (choice == 2) {
                std::wcout << L"*Изменения были утеряны\n";
            }
            else if (choice == 1) {
                InputFileName(c);
                c.save_to_file();
            }
            else if (choice == 0) {
                c.save_to_file();
            }
        }
        else {
            std::wcout << L"*Внесенные изменения являются новыми и не сохранены.\n";
            if (GetCorrectNumber(std::cin, 0, 1, L"Хотите сохранить изменения в файле? (1 - да, 0 - нет): ") == 0) {
                std::cout << "*Данные были утеряны\n";
            }
            else {
                InputFileName(c);
                c.save_to_file();
            }
        }
    }
}

void CheckBeforeLoad(ShpaginCompany& pts)
{
    CheckBeforeExit(pts);
    InputFileName(pts);
}

File: CSharp/Menu.cs
using System;

namespace CSharp
{
    internal static class Menu
    {
        private static void BackToMenu()
        {
            Console.Write("\n\nПожалуйста, нажмите Enter, чтобы вернуться в меню...");
            while (Console.ReadKey(true).Key != ConsoleKey.Enter)
            {
                continue;
            }
            Console.Clear();
        }

        public static void MainMenu(ShpaginCompany company)
        {
            string main_menu = $"1. Добавить сотрудника\n" +
                                "2. Добавить разработчика\n" +
                                "3. Просмотреть всех\n" +
                                "4. Сохранить\n" +
                                "5. Загрузить\n" +
                                "6. Очистить\n" +
                                "0. Выход";
            Console.WriteLine(main_menu);
            int menu = Utils.GetCorrectIntNumber(">> ", 0, 6);
            Console.Clear();
            switch (menu)
            {
                case 1:
                    {
                        Console.WriteLine("-> Добавить сотрудника");
                        company.add_employee();
                        break;
                    }
                case 2:
                    {
                        Console.WriteLine("-> Добавить разработчика");
                        company.add_developer();
                        break;
                    }
                case 3:
                    {
                        Console.WriteLine("-> Просмотреть всех");
                        company.console_output();
                        break;
                    }
                case 4:
                    {
                        Console.WriteLine("-> Сохранить");
                        try
                        {
                            if (CheckBeforeSave(company))
                            {
                                company.save_to_file();
                            }
                        }
                        catch (Exception)
                        {
                            Console.WriteLine("**Ошибка сохранения файла");
                        }
                        break;
                    }
                case 5:
                    {
                        Console.WriteLine("-> Загрузить");
                        try
                        {
                            CheckBeforeLoad(company);
                            company.load_from_file();
                        }
                        catch (Exception)
                        {
                            Console.WriteLine("**Ошибка загрузки файла");
                        }
                        break;
                    }
                case 6:
                    {
                        Console.WriteLine("-> Очистить");
                        company.clear_company();
                        break;
                    }
                case 0:
                default:
                    {
                        CheckBeforeExit(company);
                        Environment.Exit(0);
                        break;
                    }
            }
            BackToMenu();
        }
        private static void InputFileName(ShpaginCompany c)
        {
            if (c.show_saves())
            {
                string fn = Utils.GetCorrectString("Введите имя файла: ");
                c.FileName = fn;
            }
            else
            {
                throw new TaskCanceledException();
            }
        }

        private static bool CheckBeforeSave(ShpaginCompany c)
        {
            bool HandleSaved()
            {
                Console.WriteLine("*В компании нет несохраненных изменений");
                return false;
            }

            bool HandleEmpty()
            {
                Console.WriteLine("*В компании нечего сохранять");
                return false;
            }

            bool HandleUnsavedWithFile()
            {
                switch (Utils.GetCorrectIntNumber("Хотите сохранить в том же файле? (0 - тот же, 1 - другой, 2 - не сохранять): ", 0, 2))
                {
                    case 2:
                        Console.WriteLine("*Изменения были утеряны");
                        return false;
                    case 1:
                        InputFileName(c);
                        break;
                }
                return true;
            }

            bool HandleUnsavedWithoutFile()
            {
                if (c.Empty)
                {
                    return HandleEmpty();
                }
                InputFileName(c);
                return true;
            }

            if (c.Saved)
            {
                return c.Empty ? HandleEmpty() : HandleSaved();
            }
            return c.HasSavedFile ? HandleUnsavedWithFile() : HandleUnsavedWithoutFile();
        }
        private static void CheckBeforeExit(ShpaginCompany c)
        {
            if (!c.Saved)
            {
                if (Utils.GetCorrectIntNumber("Хотите сохранить изменения в файле? (0 - нет, 1 - да): ", 0, 1) == 1)
                {
                    c.save_to_file();
                }
            }
        }

        private static void CheckBeforeLoad(ShpaginCompany c)
        {
            CheckBeforeExit(c);
            InputFileName(c);
        }
    }
}
